### PR TITLE
ISPN-13261 Server parser and configuration refactoring

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-13261
https://issues.redhat.com/browse/ISPN-13284
https://issues.redhat.com/browse/ISPN-13283

This PR refactors the Server parser and configuration to achieve the following:
* construction of the security realms and implicit setup of the endpoints/connectors happens after parsing is complete, not during
* the parser now supports unordered elements which should help out with YAML/JSON configurations which have been processed by external tools
* Resolution of paths is done after parsing and `relative-to` property names are retained
* Bug fixes to the schema and some of the realms (mostly LDAP)
* Upgrade to Elytron 1.17.0.Final
* Use the Infinispan timeout executor for the Elytron SecurityDomain

Bonus points for removing more lines than are being added :)

### CODE REVIEW

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/ConfigurationElement.java [hunk 1]
```diff
@@ -38,7 +38,7 @@ protected ConfigurationElement(String element, boolean repeated, AttributeSet at
       this.children = children != null ? children : CHILDLESS;
    }
 
-   public String name() {
+   public final String elementName() {
```
Comment (ryanemerson) [inline commons/all/src/main/java/org/infinispan/commons/configuration/attributes/ConfigurationElement.java:41]: Any reason for this change? IMO `ConfigurationElement#ElementName` seems  unnecessarily verbose compared to `ConfigurationElement#name`.

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/NamedSocketFactory.java [hunk 2]
```diff
@@ -23,13 +24,18 @@
...
-   private final javax.net.SocketFactory socketFactory;
-   private final ServerSocketFactory serverSocketFactory;
+   private final Supplier<javax.net.SocketFactory> socketFactory;
+   private final Supplier<ServerSocketFactory> serverSocketFactory;
    private String name;
    private BiConsumer<String, Socket> socketConfigurator = (c, s) -> {};
    private BiConsumer<String, ServerSocket> serverSocketConfigurator = (c, s) -> {};
 
    public NamedSocketFactory(javax.net.SocketFactory socketFactory, ServerSocketFactory serverSocketFactory) {
+      this.socketFactory = () -> socketFactory;
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/remoting/transport/jgroups/NamedSocketFactory.java:34]: `setSocketConfigurator` and `setServerSocketConfigurator` can also be removed.

File: server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationBuilder.java [hunk 3]
```diff
@@ -21,16 +25,29 @@
 public class ServerConfigurationBuilder implements Builder<ServerConfiguration> {
    private final GlobalConfigurationBuilder builder;
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationBuilder.java:26]: unused var. Can be removed with constructor.

File: server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationBuilder.java [hunk 4]
```diff
@@ -116,4 +109,32 @@ public void applySocketBinding(String bindingName, ProtocolServerConfigurationBu
...
+   public Supplier<SSLContext> clientSSLContextSupplier(String sslContextName) {
+      SSLContextSupplier supplier = new SSLContextSupplier(sslContextName, true);
+      suppliers.add(supplier);
+      return supplier;
+   }
+
+   private static class SSLContextSupplier implements Supplier<SSLContext> {
+      final private String name;
+      private final boolean client;
+      private SecurityConfiguration configuration;
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationBuilder.java:128]: Nitpick. No need for `private` modifiers on vars.

File: server/runtime/src/main/java/org/infinispan/server/configuration/hotrod/HotRodServerConfigurationParser.java [hunk 5]
```diff
@@ -153,8 +150,8 @@ private void parseEncryption(ConfigurationReader reader, ServerConfigurationBuil
       if (securityRealm == null) {
          throw Server.log.encryptionWithoutSecurityRealm();
       } else {
-         String name = securityRealm.getName();
-         encryption.realm(name).sslContext(serverBuilder.getSSLContext(name));
+         String name = securityRealm;
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/configuration/hotrod/HotRodServerConfigurationParser.java:153]: Redundant var.

File: server/runtime/src/main/java/org/infinispan/server/configuration/security/CredentialStoresConfiguration.java [hunk 6]
```diff
@@ -1,30 +1,68 @@
...
+import org.wildfly.security.credential.Credential;
+import org.wildfly.security.credential.PasswordCredential;
+import org.wildfly.security.password.interfaces.ClearPassword;
 
 /**
  * @author Tristan Tarrant &lt;tristan@infinispan.org&gt;
  * @since 12.0
  **/
 public class CredentialStoresConfiguration extends ConfigurationElement<CredentialStoresConfiguration> {
    private final AttributeSet attributes;
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/configuration/security/CredentialStoresConfiguration.java:19]: Never used.

File: server/runtime/src/main/java/org/infinispan/server/configuration/security/CredentialStoresConfigurationBuilder.java [hunk 7]
```diff
@@ -30,41 +33,44 @@ public CredentialStoreConfigurationBuilder addCredentialStore(String name) {
...
+      return credentialSupplier;
    }
 
-   @Override
-   public Builder<?> read(CredentialStoresConfiguration template) {
-      return this;
+   private static class CredentialSupplier implements Supplier<char[]> {
+      final private String store;
+      final private String alias;
+      private CredentialStoresConfiguration configuration;
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/configuration/security/CredentialStoresConfigurationBuilder.java:63]: `private` modifier on vars and constructor is redundant.

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeSerializer.java [hunk 8]
```diff
@@ -25,6 +26,13 @@
          writer.writeAttribute(name, "***");
       }
    };
+   AttributeSerializer<Supplier<char[]>> SECRET_SUPPLIER = (writer, name, value) -> {
+      if (Boolean.getBoolean("org.infinispan.configuration.clear-text-secrets")) {
+         writer.writeAttribute(name, new String(value.get()));
+      } else {
+         writer.writeAttribute(name, "***");
+      }
+   };
```
Comment (pruivo) [inline commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeSerializer.java:35]:
```suggestion
   AttributeSerializer<Supplier<char[]>> SECRET_SUPPLIER = (writer, name, value) -> SECRET.serialize(writer, name, value.get());
```

File: core/src/main/java/org/infinispan/configuration/cache/IndexMergeConfigurationBuilder.java [hunk 9]
```diff
@@ -93,10 +93,6 @@ public String toString() {
             '}';
    }
 
-   @Override
-   public void validate() {
-   }
-
    @Override
    public void validate(GlobalConfiguration globalConfig) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/cache/IndexMergeConfigurationBuilder.java:97]: nit remove too?

File: core/src/main/java/org/infinispan/configuration/cache/TakeOfflineConfigurationBuilder.java [hunk 10]
```diff
@@ -81,9 +72,4 @@ public TakeOfflineConfigurationBuilder read(TakeOfflineConfiguration template) {
       this.attributes.read(template.attributes());
       return this;
    }
-
-   @Override
-   public String toString() {
-      return this.getClass().getSimpleName() + attributes;
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/cache/TakeOfflineConfigurationBuilder.java:87]: is this intentional?

File: server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfiguration.java [hunk 11]
```diff
@@ -80,13 +74,103 @@ public String name() {
...
+      } else {
+         return clientSslContext;
+      }
+   }
+
+   public boolean hasSslIdentity() {
+      return serverIdentitiesConfiguration.sslConfiguration() != null;
+   }
+
+   ServerSecurityRealm init(SecurityConfiguration security, Properties properties) {
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/configuration/security/RealmConfiguration.java:109]: Return value is never used.

File: core/src/main/java/org/infinispan/configuration/cache/TakeOfflineConfigurationBuilder.java (hunk 10 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/configuration/cache/TakeOfflineConfigurationBuilder.java:87]: As I was going through removing all of the empty `validate()` methods, I also got carried away removing useless method overrides on the builder objects (`hashCode`/`equals`/`toString`) unless strictly necessary

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/ConfigurationElement.java (hunk 1 above)
Comment (tristantarrant) [reply commons/all/src/main/java/org/infinispan/commons/configuration/attributes/ConfigurationElement.java:41]: Unfortunately many builders were overriding `name()` with their interpretation: we want it to mean the declarative name, not some logical value

**Comment by tristantarrant** [comment]
> It's still not possible to have `endpoints` elements proceed `socket-bindings` due to `applySocketBinding` being called from the `EndpointConfigurationBuilder` constructor.

Let me fix that

File: server/runtime/src/main/java/org/infinispan/server/configuration/hotrod/HotRodServerConfigurationParser.java [hunk 12]
```diff
@@ -67,7 +66,7 @@ public void readElement(ConfigurationReader reader, ConfigurationBuilderHolder h
 
    private void parseHotRodConnector(ConfigurationReader reader, ConfigurationBuilderHolder holder, ServerConfigurationBuilder serverBuilder, HotRodServerConfigurationBuilder builder) {
       boolean dedicatedSocketBinding = false;
-      ServerSecurityRealm securityRealm = null;
+      String securityRealm = null;
```
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/configuration/hotrod/HotRodServerConfigurationParser.java:69]: `holder` var is never used.

File: server/core/src/main/java/org/infinispan/server/core/configuration/SslEngineConfigurationBuilder.java [hunk 13]
```diff
@@ -125,7 +136,7 @@ public void validate() {
       if(domain == null) {
          throw log.noSniDomainConfigured();
       }
-      if (sslContext == null) {
+      if (sslContextSupplier == null) {
```
Comment (pruivo) [inline server/core/src/main/java/org/infinispan/server/core/configuration/SslEngineConfigurationBuilder.java:139]: missing checking the supplier
```suggestion
      if (sslContextSupplier == null || sslContextSupplier.get() == null) {
```

File: server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java [hunk 14]
```diff
@@ -736,27 +734,27 @@ private void parseLdapIdentityMapping(ConfigurationReader reader, LdapIdentityMa
             case SEARCH_RECURSIVE:
                identityMapBuilder.searchRecursive(Boolean.valueOf(value));
                break;
+            case SEARCH_TIME_LIMIT:
+               identityMapBuilder.searchTimeLimit(Integer.valueOf(value));
```
Comment (pruivo) [inline server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java:738]: avoid boxing&unboxing
```suggestion
               identityMapBuilder.searchTimeLimit(Integer.parseInt(value));
```

**Review by pruivo — CHANGES_REQUESTED** [review]

File: server/core/src/main/java/org/infinispan/server/core/configuration/SslEngineConfigurationBuilder.java (hunk 13 above)
Comment (tristantarrant) [reply server/core/src/main/java/org/infinispan/server/core/configuration/SslEngineConfigurationBuilder.java]: Fixed

File: server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java (hunk 14 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java]: Fixed

File: server/runtime/src/main/java/org/infinispan/server/configuration/hotrod/HotRodServerConfigurationParser.java (hunk 12 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/configuration/hotrod/HotRodServerConfigurationParser.java:69]: Fixed

File: commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java [hunk 15]
```diff
@@ -180,7 +196,7 @@ Parsed parseLine(final String s) {
             int c = s.charAt(i);
             switch (c) {
                case ' ':
-                  if (state == 0 && !parsed.listItem) {
+                  if (state == 0) {// && !parsed.listItem) {
```
Comment (ryanemerson) [inline commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java:199]: Comment can be removed.

File: commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java [hunk 16]
```diff
@@ -44,6 +44,21 @@ public YamlConfigurationReader(Reader reader, ConfigurationResourceResolver reso
...
+         printTree(lines);
+      }
+   }
+
+   private void printTree(Node node) {
+      if (node != null) {
+         for (int i = 0; i < node.parsed.indent; i++) {
+            System.out.print(' ');
+         }
+         System.out.println(node.parsed);
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java]: Done

File: commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java (hunk 15 above)
Comment (tristantarrant) [reply commons/all/src/main/java/org/infinispan/commons/configuration/io/yaml/YamlConfigurationReader.java]: Done

**Comment by ryanemerson** [comment]
`YamlConfigurationReaderTest` failure fixed since last build ... merging.

Thanks @tristantarrant

**Pull request merged**